docs/auth.md

```diff
@@ -0,0 +1,96 @@
+
+The auth system consists of:
+
+- Users
+- Permissions: Binary (yes/no) flags designating whether a user may perform a certain task.
+- Default Permissions
+- Roles: A generic way of applying labels and permissions to more than one user.
+- A password hashing system
+- APIs for logging in users or restricting content
+
+The authentication system in Soul aims to be very generic and doesn't provide some features commonly found in web authentication systems, such as:
+
+- Password strength checking
+- Throttling of login attempts
+- Authentication against third-parties (OAuth, for example)
+- Object-level permissions
+
+#### Users
+
+User objects are the core of the authentication system. They typically represent the people interacting with your site and are used to enable things like restricting access, registering user profiles, associating content with creators, etc. Only one class of user exists in Soul's authentication framework, i.e., 'superusers' or types of users are just user objects with special attributes set, not different classes of user objects.
+
+The primary attributes of the default user are:
+
+- username
+- password
+- email
+- first_name
+- last_name
+
+Note that when Soul boots up, it looks for a table called "\_users" which holds the User objects mentioned above.
+
+Creating superusers:
+
+Create superusers using the createsuperuser command:
+
+```
+$ node src/server.js createsuperuser --username=joe --password=strongstring
+```
+
+#### Permissions
+
+Soul comes with a built-in permissions system. It provides a way to assign permissions to specific users and groups of users.
+
+Soul uses permissions as follows:
+
+- Access to view objects is limited to users with the "read" permission for that type of object.
+- Access to add an object is limited to users with the "create" permission for that type of object.
+- Access to change an object is limited to users with the "update" permission for that type of object.
+- Access to delete an object is limited to users with the "delete" permission for that type of object.
+
+Default Permissions:
+
+By default, Soul ensures that four default permissions – create, update, delete, and read – are created for each table defined in one of your databases.
+
+| table_name | can_create | can_read | can_update | can_delete |
+| ---------- | ---------- | -------- | ---------- | ---------- |
+| albums     | false      | true     | true       | false      |
+
+In this scenario, calling POST (create) and GET (read) requests to /tables/albums/rows is allowed by any user (even anonymous), while PUT (update) and DELETE (delete, of course) requests are not allowed.
+
+When Soul boots up, it checks for the existence of a table called \_default_permissions.
+
```

Comment:
The `_default_permissions` and `_permissions` should be automatically created during the initialization of Soul or whenever a new table is created.

Reply:
> The `_default_permissions` and `_permissions` should be automatically created during the initialization of Soul or whenever a new table is created.

Would Soul create these?  Or is it the responsibility of a soul instance?

Reply:
Soul will create these:

1. Once booting up
2. Once a new table is created

Reply:
Thanks @thevahidal - could you include that in the document please?  It will be _lost_ in here, but will live forever in `auth.md` :-D

Reply:
Sure thing!

Reply:
The @thevahidal - I think we need to expand on the `default role` permissions.  You indicate that we look for `default` role when a new table is created.

I think there is another perspective.  I think that when we create a new role, we look for a `default` role, and duplicate the table-based `role_permissions` from `default` for the new `role`.

In the absence of a row in `role_permissions` for the `default` role for that `table` we generate `role_permissions` to provide just read access.

So, from the above, the `default` `role` is used for creating permissions for a new role, not for creating a new table.  

Aah, I've got it.

>  then looks for existence of role permission with role of default for the table

There won't be an existing default for the table, since it's a new table.

Reply:
I see why this might be confusing, I rewrite it so it becomes more clear. 
Let me know what you think.